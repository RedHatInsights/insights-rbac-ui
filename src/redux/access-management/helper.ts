import { getAccessApi } from '../../api/httpClient';

// Interface for helper parameters - exported for reuse in actions
export interface GetPrincipalAccessParams {
  limit?: number;
  offset?: number;
  username?: string;
  application?: string; // Often comma-separated list
  status?: string;
  orderBy?: string;
}

const accessApi = getAccessApi();

export async function getPrincipalAccess({ limit, offset, username, application = '', status, orderBy }: GetPrincipalAccessParams): Promise<unknown> {
  // NOTE: @redhat-cloud-services/rbac-client has completely broken TypeScript types
  // - API expects undefined for optional parameters, not explicit defaults
  // - Using (apiMethod as any) to bypass broken type definitions
  return await (accessApi.getPrincipalAccess as any)(application, username, orderBy, status, limit, offset);
}
