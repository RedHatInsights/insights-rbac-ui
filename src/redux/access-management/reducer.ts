import { GET_PRINCIPAL_ACCESS } from './action-types';
import { PaginationDefaultI, defaultSettings } from '../../helpers/pagination';

// Type definitions for access management
interface AccessItem {
  permission: string;
  application: string;
  resource_type: string;
  verb: string;
  resourceDefinitions?: unknown[];
  [key: string]: unknown;
}

interface AccessData {
  data: AccessItem[];
  meta: PaginationDefaultI;
}

export interface AccessState {
  isLoading: boolean;
  access: AccessData;
  [key: string]: unknown;
}

// Using global ReduxAction interface from store.d.ts instead of local Action interface

// Initial State
export const accessInitialState: AccessState = {
  isLoading: false,
  access: {
    data: [],
    meta: defaultSettings,
  },
};

const setLoadingState = (state: AccessState): AccessState => ({ ...state, isLoading: true });

const setAccess = (state: AccessState, action?: ReduxAction<AccessData>): AccessState => {
  if (!action) return state;
  const { payload } = action;
  return { ...state, access: payload, isLoading: false };
};

// Reducer function map with proper typing
type ReducerFunctions = Record<string, (state: AccessState, action?: ReduxAction<any>) => AccessState>;

const accessReducer: ReducerFunctions = {
  [`${GET_PRINCIPAL_ACCESS}_PENDING`]: setLoadingState,
  [`${GET_PRINCIPAL_ACCESS}_FULFILLED`]: setAccess,
};

export default accessReducer;
