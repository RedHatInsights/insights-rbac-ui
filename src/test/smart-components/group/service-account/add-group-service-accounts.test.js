import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { render, screen } from '@testing-library/react';
import * as ServiceAccountsHelper from '../../../../redux/service-accounts/helper';
import AddGroupServiceAccounts from '../../../../features/groups/service-account/add-group-service-accounts';
import { serviceAccountsInitialState } from '../../../../redux/service-accounts/reducer';
import * as useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { RESULTS } from '../../../../redux/service-accounts/constants';

jest.mock('@redhat-cloud-services/frontend-components/useChrome', () => {
  return {
    __esModule: true,
    default: () => ({}),
    useChrome: () => {},
  };
});

describe('<AddGroupServiceAccounts />', () => {
  let initialState;
  let responseBody;
  let mockStore;

  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  const fetchServiceAccountsSpy = jest.spyOn(ServiceAccountsHelper, 'getServiceAccounts');

  beforeEach(() => {
    initialState = {
      groupReducer: {
        groups: {
          data: [
            {
              uuid: '123',
              name: 'SampleGroup',
            },
          ],
        },
        systemGroup: {
          uuid: 'test',
        },
      },
      serviceAccountReducer: {
        ...serviceAccountsInitialState,
      },
    };
    responseBody = {
      data: [],
      limit: 1,
      offset: 0,
      status: RESULTS,
    };
    mockStore = configureStore(middlewares);
  });

  afterEach(() => {
    fetchServiceAccountsSpy.mockReset();
  });

  test('should render correctly empty', () => {
    jest.spyOn(useChrome, 'default').mockImplementationOnce(() => ({
      getEnvironmentDetails: () => {},
      auth: { getToken: () => 'token' },
    }));
    fetchServiceAccountsSpy.mockImplementationOnce(() => Promise.resolve(responseBody));
    const store = mockStore({ ...initialState });
    const container = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/service-accounts/add-service-account']}>
          <AddGroupServiceAccounts />
        </MemoryRouter>
      </Provider>,
    );
    expect(container.baseElement).toMatchSnapshot();
    expect(screen.getByText('No service accounts found')).toBeInTheDocument();
  });

  test('should render correctly with data', () => {
    jest.spyOn(useChrome, 'default').mockImplementationOnce(() => ({
      getEnvironmentDetails: () => {},
      auth: { getToken: () => 'token' },
    }));
    fetchServiceAccountsSpy.mockImplementationOnce(() => Promise.resolve(responseBody));
    const store = mockStore({
      ...initialState,
      serviceAccountReducer: {
        ...initialState.serviceAccountReducer,
        serviceAccounts: [
          {
            clientId: '0f81d08d-74b2',
            createdAt: 1705679691,
            createdBy: 'insights-qa',
            description: 'just-created',
            id: '0f81d08d-74b2',
            name: 'just-created',
          },
        ],
      },
    });
    const container = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/service-accounts/add-service-account']}>
          <AddGroupServiceAccounts />
        </MemoryRouter>
      </Provider>,
    );
    expect(container.baseElement).toMatchSnapshot();
    expect(screen.getAllByText('just-created')).toHaveLength(2);
  });
});
