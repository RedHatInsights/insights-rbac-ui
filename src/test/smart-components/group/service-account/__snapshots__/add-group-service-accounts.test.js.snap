// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddGroupServiceAccounts /> should render correctly empty 1`] = `
<body
  class="pf-v5-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div
    class="pf-v5-c-backdrop"
    id="pf-modal-part-3"
  >
    <div
      class="pf-v5-l-bullseye"
    >
      <div
        aria-describedby="pf-modal-part-2"
        aria-labelledby="pf-modal-part-1"
        aria-modal="true"
        class="pf-v5-c-modal-box rbac pf-m-md"
        data-ouia-component-id="OUIA-Generated-Modal-medium-1"
        data-ouia-component-type="PF5/ModalContent"
        data-ouia-safe="true"
        id="pf-modal-part-0"
        role="dialog"
      >
        <div
          class="pf-v5-c-modal-box__close"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-v5-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-medium-1-ModalBoxCloseButton"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
        </div>
        <header
          class="pf-v5-c-modal-box__header"
        >
          <h1
            class="pf-v5-c-modal-box__title"
            id="pf-modal-part-1"
          >
            <span
              class="pf-v5-c-modal-box__title-text"
            >
              Add service account
            </span>
          </h1>
        </header>
        <div
          class="pf-v5-c-modal-box__body"
          id="pf-modal-part-2"
        >
          <div
            class="pf-v5-l-stack pf-m-gutter"
          >
            <div
              class="pf-v5-l-stack__item"
            >
              <div
                class="pf-v5-c-content"
              >
                This list contains all service accounts associated with your Red Hat organization account. Select any service accounts you wish to associate with the User Access group.
                <div
                  class="pf-v5-c-alert pf-m-inline pf-m-plain pf-m-info pf-v5-u-mt-sm rbac-service-accounts-alert"
                  data-ouia-component-id="OUIA-Generated-Alert-info-1"
                  data-ouia-component-type="PF5/Alert"
                  data-ouia-safe="true"
                >
                  <div
                    class="pf-v5-c-alert__icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                      />
                    </svg>
                  </div>
                  <span
                    class="pf-v5-c-alert__title"
                  >
                    <span
                      class="pf-v5-screen-reader"
                    >
                      Info alert:
                    </span>
                    To add, reset credentials, or delete service accounts visit the {link}.
                  </span>
                </div>
              </div>
            </div>
            <div
              class="pf-v5-l-stack__item rbac-add-service-account-modal"
            >
              <div
                class="rbac-service-accounts-list"
              >
                <div
                  class="pf-v5-c-empty-state pf-m-sm"
                >
                  <div
                    class="pf-v5-c-empty-state__content"
                  >
                    <div
                      class="pf-v5-c-empty-state__header"
                    >
                      <div
                        class="pf-v5-c-empty-state__icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                          />
                        </svg>
                      </div>
                      <div
                        class="pf-v5-c-empty-state__title"
                      >
                        <h4
                          class="pf-v5-c-empty-state__title-text"
                        >
                          No service accounts found
                        </h4>
                      </div>
                    </div>
                    <div
                      class="pf-v5-c-empty-state__body pf-v5-u-mb-md"
                    >
                      Contact your platform service team to add service accounts.
                       
                      <br />
                       
                      <br />
                    </div>
                    <div
                      class="pf-v5-c-empty-state__footer"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="pf-v5-c-modal-box__footer"
        >
          <button
            aria-disabled="true"
            class="pf-v5-c-button pf-m-primary pf-m-disabled"
            data-ouia-component-id="primary-confirm-button"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            disabled=""
            type="button"
          >
            Add to group
          </button>
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-link"
            data-ouia-component-id="secondary-cancel-button"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            Cancel
          </button>
        </footer>
      </div>
    </div>
  </div>
</body>
`;

exports[`<AddGroupServiceAccounts /> should render correctly with data 1`] = `
<body
  class="pf-v5-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div
    class="pf-v5-c-backdrop"
    id="pf-modal-part-4"
  >
    <div
      class="pf-v5-l-bullseye"
    >
      <div
        aria-describedby="pf-modal-part-3"
        aria-labelledby="pf-modal-part-2"
        aria-modal="true"
        class="pf-v5-c-modal-box rbac pf-m-md"
        data-ouia-component-id="OUIA-Generated-Modal-medium-2"
        data-ouia-component-type="PF5/ModalContent"
        data-ouia-safe="true"
        id="pf-modal-part-1"
        role="dialog"
      >
        <div
          class="pf-v5-c-modal-box__close"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-v5-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-medium-2-ModalBoxCloseButton"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
        </div>
        <header
          class="pf-v5-c-modal-box__header"
        >
          <h1
            class="pf-v5-c-modal-box__title"
            id="pf-modal-part-2"
          >
            <span
              class="pf-v5-c-modal-box__title-text"
            >
              Add service account
            </span>
          </h1>
        </header>
        <div
          class="pf-v5-c-modal-box__body"
          id="pf-modal-part-3"
        >
          <div
            class="pf-v5-l-stack pf-m-gutter"
          >
            <div
              class="pf-v5-l-stack__item"
            >
              <div
                class="pf-v5-c-content"
              >
                This list contains all service accounts associated with your Red Hat organization account. Select any service accounts you wish to associate with the User Access group.
                <div
                  class="pf-v5-c-alert pf-m-inline pf-m-plain pf-m-info pf-v5-u-mt-sm rbac-service-accounts-alert"
                  data-ouia-component-id="OUIA-Generated-Alert-info-2"
                  data-ouia-component-type="PF5/Alert"
                  data-ouia-safe="true"
                >
                  <div
                    class="pf-v5-c-alert__icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                      />
                    </svg>
                  </div>
                  <span
                    class="pf-v5-c-alert__title"
                  >
                    <span
                      class="pf-v5-screen-reader"
                    >
                      Info alert:
                    </span>
                    To add, reset credentials, or delete service accounts visit the {link}.
                  </span>
                </div>
              </div>
            </div>
            <div
              class="pf-v5-l-stack__item rbac-add-service-account-modal"
            >
              <div
                class="rbac-service-accounts-list"
              >
                <div
                  class="pf-v5-c-toolbar  ins-c-primary-toolbar"
                  data-ouia-component-id="PrimaryToolbar"
                  data-ouia-component-type="PF5/Toolbar"
                  data-ouia-safe="true"
                  id="ins-primary-data-toolbar"
                >
                  <div
                    class="pf-v5-c-toolbar__content"
                  >
                    <div
                      class="pf-v5-c-toolbar__content-section"
                    >
                      <div
                        class="pf-v5-c-toolbar__group pf-m-filter-group ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
                      >
                        <div
                          class="pf-v5-c-toolbar__item"
                        >
                          <button
                            aria-expanded="false"
                            class="pf-v5-c-menu-toggle"
                            data-ouia-component-id="OUIA-Generated-MenuToggle-1"
                            data-ouia-component-type="PF5/MenuToggle"
                            data-ouia-safe="true"
                            type="button"
                          >
                            <span
                              class="pf-v5-c-menu-toggle__text"
                            >
                              <label
                                class="pf-v5-c-check pf-m-standalone"
                              >
                                <input
                                  aria-invalid="false"
                                  aria-label="Select all"
                                  class="pf-v5-c-check__input"
                                  data-ouia-component-id="BulkSelectCheckbox"
                                  data-ouia-component-type="PF5/MenuToggleCheckbox"
                                  data-ouia-safe="true"
                                  name="group-add-accounts-toggle-checkbox"
                                  type="checkbox"
                                />
                              </label>
                            </span>
                            <span
                              class="pf-v5-c-menu-toggle__controls"
                            >
                              <span
                                class="pf-v5-c-menu-toggle__toggle-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="pf-v5-svg"
                                  fill="currentColor"
                                  height="1em"
                                  role="img"
                                  viewBox="0 0 320 512"
                                  width="1em"
                                >
                                  <path
                                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </button>
                        </div>
                        <div
                          class="pf-v5-c-toolbar__item ins-c-primary-toolbar__filter"
                        >
                          <div
                            class="pf-v5-l-split ins-c-conditional-filter"
                          >
                            <div
                              class="pf-v5-l-split__item pf-m-fill"
                              data-ouia-component-id="ConditionalFilter"
                            >
                              <span
                                class="pf-v5-c-form-control pf-m-icon ins-c-conditional-filter"
                              >
                                <input
                                  aria-invalid="false"
                                  aria-label="text input"
                                  data-ouia-component-id="ConditionalFilter"
                                  data-ouia-component-type="PF5/TextInput"
                                  data-ouia-safe="true"
                                  id="filter-by-string"
                                  placeholder="Filter by {key}"
                                  type="text"
                                  value=""
                                  widget-type="InsightsInput"
                                />
                                <span
                                  class="pf-v5-c-form-control__utilities"
                                >
                                  <span
                                    class="pf-v5-c-form-control__icon"
                                  >
                                    <span
                                      class="pf-v5-c-icon pf-m-md ins-c-search-icon"
                                    >
                                      <span
                                        class="pf-v5-c-icon__content"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          class="pf-v5-svg"
                                          fill="currentColor"
                                          height="1em"
                                          role="img"
                                          viewBox="0 0 512 512"
                                          width="1em"
                                        >
                                          <path
                                            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        class="pf-v5-c-toolbar__item ins-c-primary-toolbar__pagination"
                      >
                        <div
                          class="pf-v5-c-pagination pf-m-compact"
                          data-ouia-component-id="CompactPagination"
                          data-ouia-component-type="PF5/Pagination"
                          data-ouia-safe="true"
                          id="options-menu-top-pagination"
                          style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
                        >
                          <div
                            class="pf-v5-c-pagination__total-items"
                          >
                            <b>
                              1
                               - 
                              1
                            </b>
                             
                            of 
                            <b>
                              many
                            </b>
                          </div>
                          <button
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
                            data-ouia-component-id="OUIA-Generated-MenuToggle-plainText-1"
                            data-ouia-component-type="PF5/MenuToggle"
                            data-ouia-safe="true"
                            id="options-menu-top-toggle"
                            type="button"
                          >
                            <span
                              class="pf-v5-c-menu-toggle__text"
                            >
                              <b>
                                1
                                 - 
                                1
                              </b>
                               
                              of 
                              <b>
                                many
                              </b>
                            </span>
                            <span
                              class="pf-v5-c-menu-toggle__controls"
                            >
                              <span
                                class="pf-v5-c-menu-toggle__toggle-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="pf-v5-svg"
                                  fill="currentColor"
                                  height="1em"
                                  role="img"
                                  viewBox="0 0 320 512"
                                  width="1em"
                                >
                                  <path
                                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </button>
                          <nav
                            aria-label="Pagination"
                            class="pf-v5-c-pagination__nav"
                          >
                            <div
                              class="pf-v5-c-pagination__nav-control"
                            >
                              <button
                                aria-disabled="true"
                                aria-label="Go to previous page"
                                class="pf-v5-c-button pf-m-plain pf-m-disabled"
                                data-action="previous"
                                data-ouia-component-id="OUIA-Generated-Button-plain-3"
                                data-ouia-component-type="PF5/Button"
                                data-ouia-safe="true"
                                disabled=""
                                type="button"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="pf-v5-svg"
                                  fill="currentColor"
                                  height="1em"
                                  role="img"
                                  viewBox="0 0 256 512"
                                  width="1em"
                                >
                                  <path
                                    d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                                  />
                                </svg>
                              </button>
                            </div>
                            <div
                              class="pf-v5-c-pagination__nav-control"
                            >
                              <button
                                aria-disabled="false"
                                aria-label="Go to next page"
                                class="pf-v5-c-button pf-m-plain"
                                data-action="next"
                                data-ouia-component-id="OUIA-Generated-Button-plain-4"
                                data-ouia-component-type="PF5/Button"
                                data-ouia-safe="true"
                                type="button"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="pf-v5-svg"
                                  fill="currentColor"
                                  height="1em"
                                  role="img"
                                  viewBox="0 0 256 512"
                                  width="1em"
                                >
                                  <path
                                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                                  />
                                </svg>
                              </button>
                            </div>
                          </nav>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="pf-v5-c-toolbar__content pf-m-hidden"
                    hidden=""
                  >
                    <div
                      class="pf-v5-c-toolbar__group"
                    />
                  </div>
                </div>
                <table
                  aria-label="service accounts table"
                  class="pf-v5-c-table pf-m-grid-md"
                  data-ouia-component-id="group-add-accounts"
                  data-ouia-component-type="PF5/Table"
                  data-ouia-safe="true"
                  role="grid"
                >
                  <thead
                    class="pf-v5-c-table__thead"
                  >
                    <tr
                      class="pf-v5-c-table__tr"
                      data-ouia-component-id="OUIA-Generated-TableRow-1"
                      data-ouia-component-type="PF5/TableRow"
                      data-ouia-safe="true"
                    >
                      <th
                        class="pf-v5-c-table__th"
                        data-key="0"
                        data-label=""
                        tabindex="-1"
                      />
                      <th
                        class="pf-v5-c-table__th"
                        data-key="1"
                        data-label="Name"
                        scope="col"
                        tabindex="-1"
                      >
                        Name
                      </th>
                      <th
                        class="pf-v5-c-table__th"
                        data-key="2"
                        data-label="Description"
                        scope="col"
                        tabindex="-1"
                      >
                        Description
                      </th>
                      <th
                        class="pf-v5-c-table__th"
                        data-key="3"
                        data-label="Client ID"
                        scope="col"
                        tabindex="-1"
                      >
                        Client ID
                      </th>
                      <th
                        class="pf-v5-c-table__th"
                        data-key="4"
                        data-label="Owner"
                        scope="col"
                        tabindex="-1"
                      >
                        Owner
                      </th>
                      <th
                        class="pf-v5-c-table__th"
                        data-key="5"
                        data-label="Time created"
                        scope="col"
                        tabindex="-1"
                      >
                        Time created
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="pf-v5-c-table__tbody"
                    role="rowgroup"
                  >
                    <tr
                      class="pf-v5-c-table__tr"
                      data-ouia-component-id="OUIA-Generated-TableRow-2"
                      data-ouia-component-type="PF5/TableRow"
                      data-ouia-safe="true"
                    >
                      <td
                        class="pf-v5-c-table__td pf-v5-c-table__check"
                        data-key="0"
                        tabindex="-1"
                      >
                        <label>
                          <input
                            aria-label="Select row 0"
                            name="checkrow0"
                            type="checkbox"
                          />
                        </label>
                      </td>
                      <td
                        class="pf-v5-c-table__td"
                        data-key="1"
                        data-label="Name"
                        tabindex="-1"
                      >
                        just-created
                      </td>
                      <td
                        class="pf-v5-c-table__td"
                        data-key="2"
                        data-label="Description"
                        tabindex="-1"
                      >
                        just-created
                      </td>
                      <td
                        class="pf-v5-c-table__td"
                        data-key="3"
                        data-label="Client ID"
                        tabindex="-1"
                      >
                        0f81d08d-74b2
                      </td>
                      <td
                        class="pf-v5-c-table__td"
                        data-key="4"
                        data-label="Owner"
                        tabindex="-1"
                      >
                        insights-qa
                      </td>
                      <td
                        class="pf-v5-c-table__td"
                        data-key="5"
                        data-label="Time created"
                        tabindex="-1"
                      >
                        <div
                          style="display: contents;"
                        >
                          <span>
                            56 years ago
                          </span>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
                <div
                  class="pf-v5-c-toolbar ins-c-table__toolbar"
                  data-ouia-component-id="OUIA-Generated-RHI/TableToolbar-true-1"
                  data-ouia-component-type="RHI/TableToolbar"
                  data-ouia-safe="true"
                  id="pf-random-id-0"
                >
                  <div
                    class="pf-v5-c-pagination pf-m-bottom pf-m-compact"
                    data-ouia-component-id="OUIA-Generated-Pagination-bottom-1"
                    data-ouia-component-type="PF5/Pagination"
                    data-ouia-safe="true"
                    id="options-menu-bottom-pagination"
                    style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
                      data-ouia-component-id="OUIA-Generated-MenuToggle-plainText-2"
                      data-ouia-component-type="PF5/MenuToggle"
                      data-ouia-safe="true"
                      id="options-menu-bottom-toggle"
                      type="button"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__text"
                      >
                        <b>
                          1
                           - 
                          1
                        </b>
                         
                        of 
                        <b>
                          many
                        </b>
                      </span>
                      <span
                        class="pf-v5-c-menu-toggle__controls"
                      >
                        <span
                          class="pf-v5-c-menu-toggle__toggle-icon"
                        >
                          <svg
                            aria-hidden="true"
                            class="pf-v5-svg"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            viewBox="0 0 320 512"
                            width="1em"
                          >
                            <path
                              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                            />
                          </svg>
                        </span>
                      </span>
                    </button>
                    <nav
                      aria-label="Pagination"
                      class="pf-v5-c-pagination__nav"
                    >
                      <div
                        class="pf-v5-c-pagination__nav-control"
                      >
                        <button
                          aria-disabled="true"
                          aria-label="Go to previous page"
                          class="pf-v5-c-button pf-m-plain pf-m-disabled"
                          data-action="previous"
                          data-ouia-component-id="OUIA-Generated-Button-plain-5"
                          data-ouia-component-type="PF5/Button"
                          data-ouia-safe="true"
                          disabled=""
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="pf-v5-svg"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            viewBox="0 0 256 512"
                            width="1em"
                          >
                            <path
                              d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                            />
                          </svg>
                        </button>
                      </div>
                      <div
                        class="pf-v5-c-pagination__nav-control"
                      >
                        <button
                          aria-disabled="false"
                          aria-label="Go to next page"
                          class="pf-v5-c-button pf-m-plain"
                          data-action="next"
                          data-ouia-component-id="OUIA-Generated-Button-plain-6"
                          data-ouia-component-type="PF5/Button"
                          data-ouia-safe="true"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="pf-v5-svg"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            viewBox="0 0 256 512"
                            width="1em"
                          >
                            <path
                              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                            />
                          </svg>
                        </button>
                      </div>
                    </nav>
                  </div>
                  <div
                    class="pf-v5-c-toolbar__content pf-m-hidden"
                    hidden=""
                  >
                    <div
                      class="pf-v5-c-toolbar__group"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="pf-v5-c-modal-box__footer"
        >
          <button
            aria-disabled="true"
            class="pf-v5-c-button pf-m-primary pf-m-disabled"
            data-ouia-component-id="primary-confirm-button"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            disabled=""
            type="button"
          >
            Add to group
          </button>
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-link"
            data-ouia-component-id="secondary-cancel-button"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            Cancel
          </button>
        </footer>
      </div>
    </div>
  </div>
</body>
`;
